ape_tenderly/provider.py

```diff
@@ -1,39 +1,113 @@
-import os
+import atexit
 
-import requests
-from ape.api import ProviderAPI, Web3Provider
-from ape.exceptions import ConfigError
+from ape.api import PluginConfig, UpstreamProvider, Web3Provider
+from ape.exceptions import ProviderError
+from ape.logging import logger
 from ape.utils import cached_property
 from web3 import HTTPProvider, Web3
+from web3.gas_strategies.rpc import rpc_gas_price_strategy
+from web3.middleware import geth_poa_middleware
 
-TENDERLY_FORK_ID = "TENDERLY_FORK_ID"
-TENDERLY_FORK_SERVICE_URI = "TENDERLY_FORK_SERVICE_URI"
+from .client import Fork, TenderlyClient
 
 
-class TenderlyProvider(Web3Provider, ProviderAPI):
+class TenderlyConfig(PluginConfig):
+    auto_remove_forks: bool = True
+
+
+class TenderlyForkProvider(Web3Provider):
+    @cached_property
+    def _client(self) -> TenderlyClient:
+        return TenderlyClient()
+
+    def _create_fork(self) -> Fork:
+        ecosystem_name = self.network.ecosystem.name
+        network_name = self.network.name.replace("-fork", "")
+        chain_id = self.network.ecosystem.get_network(network_name).chain_id
+
+        logger.debug(f"Creating tenderly fork for '{ecosystem_name}:{network_name}'...")
+        fork = self._client.create_fork(chain_id)
+        logger.success(f"Created tenderly fork '{fork.id}'.")
+        return fork
+
     @cached_property
-    def fork_id(self) -> str:
-        if TENDERLY_FORK_ID in os.environ:
-            return os.environ[TENDERLY_FORK_ID]
-
-        elif TENDERLY_FORK_SERVICE_URI in os.environ:
-            fork_network_name = self.network.name.replace("-fork", "")
-            chain_id = self.network.ecosystem.get_network(fork_network_name).chain_id
-            response = requests.post(
-                os.environ[TENDERLY_FORK_SERVICE_URI],
-                json={"network_id": str(chain_id)},
-            )
-            return response.json()["simulation_fork"]["id"]
+    def fork(self) -> Fork:
+        # NOTE: Always create a new fork, because the fork will get cached here
+        #       per-instance of this class, and "released" when the fork is closed
+        return self._create_fork()
+
+    @property
+    def uri(self) -> str:
+        return self.fork.json_rpc_url
+
+    def connect(self):
+        self._web3 = Web3(HTTPProvider(self.uri))
+        atexit.register(self.disconnect)  # NOTE: Make sure we de-provision forks
+
+    def disconnect(self):
+        if self.config.auto_remove_forks:
+            try:
+                fork_id = self.fork.id
+                logger.debug(f"Removing tenderly fork '{fork_id}'...")
+                self._client.remove_fork(fork_id)
```

Comment:
this is the only line that really needs to be in the `try` i believe

Reply:
Refactored, but kept line below as it should only appear if successfully deleted

Reply:
Fixed in [21f5a45](https://github.com/ApeWorX/ape-tenderly/pull/6/commits/21f5a4565e8a8520aed967f0c3e50bf4a8ad35f4)